### DISCUSSION TITLE
using-strongly-typed-entity-ids-to-avoid-primitive-obsession-part-1/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#73] [Announcements]
# using-strongly-typed-entity-ids-to-avoid-primitive-obsession-part-1/

In this post I show a common case where primitive GUIDs are used for entity IDs, the issues it can cause, and a possible solution using strongly-typed IDs.

https://andrewlock.net/using-strongly-typed-entity-ids-to-avoid-primitive-obsession-part-1/

**Comment by vaishnavkishan** [comment]
Would it be wise to define implicit and explicit conversion methods on the type? Because doing that covers everything like comparability, equality, etc.

For e.g. Below type is wrapper around string to allow trimming of assigning string.
```
public readonly struct Description
    {
        private readonly string _description;

        public Description(string description)
        {
            string desc = description;
            if (desc.Length > ProductConst.DescriptionLength)
                desc = description.Remove(ProductConst.DescriptionLength);

            _description = desc;
        }

        public static explicit operator Description(string value) => new Description(value);
        public static implicit operator string(Description description) => description._description;
    }
```

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at April 10 2019, 08:20, in reply to [Rival Abdrakhmanov](https://disqus.com/by/rafaelldi/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/73#discussioncomment-6056402)</em>

---

<p>A great point (that I wasn't aware of!) On the positive side (and if I'm understanding it correctly) you can safely use a readonly struct everywhere (so marking the struct readonly is safe), but you need to add the `in` modifier for maximum effectiveness. But never use the `in` modify with non-readonly structs!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at April 10 2019, 08:16, in reply to [Stilgar](https://disqus.com/by/NaibStilgar/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/73#discussioncomment-6056409)</em>

---

<p>Next post should alleviate some of those concerns too 🙂</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at April 10 2019, 09:01, in reply to [Muhammad Rehan Saeed](https://disqus.com/by/RehanSaeedUK/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/73#discussioncomment-6056422)</em>

---

<p>Potentially could inherit from Guid - I considered using a base class but don't want people to be able to accept Guids/BaseId in services in general, as loses the benefits. It's a trade off :) The boilerplate is a pain, but the reality is that you're going to generate the code, and then never touch it again, so I don't think it's a big issue personally.</p><p>Could you get the best of both worlds with the validation e.g. something like:<br><code><br>public class ZipCode<br>{<br>    public string Value { get; }<br>    public ZipCode(string input) {<br>        if(!IsValid(input)) throw new ArgumentException("input");<br><br>        Value = input;<br>    }<br><br>    static bool IsValid(string input){<br>        // do validation<br>    }<br>}<br></code></p><p>Or keep constructor private and have `TryCreate` methods etc?</p>

[a comment by an automated account is not shown]

**Reply by MattSpeakman** [reply]
Could not agree more. This is the most pointless bit of code that could be avoided by using unit tests, integration tests and/or functional tests

[a comment by an automated account is not shown]

**Comment by InteXX** [comment]
Ah, but we shouldn't be using Order.Id in our logic in the first place.

Primary and foreign keys are for use by the underlying data store for the enforcement of referential integrity only. They have no place in application logic or—heaven forbid!—the user interface (to include a browser's address/url field).

If we find ourselves using a primary or a foreign key in our code at all (such as in this example), we can pretty well be assured that we're violating a fundamental rule before we get off the ground.

Now: strongly typed _candidate_ keys... well, that's a different question entirely.

**Comment by MattSpeakman** [comment]
This is why I write unit tests and integration tests.

**Comment by mdelgadov** [comment]
I got here from a video from Milos Jovanovic. He uses Records and does not use your nuget pack. 

Any comment about that?

**Reply by andrewlock** [reply]
Not really, that's a perfectly reasonable approach 😄 I mentioned it several times [in this post](https://andrewlock.net/strongly-typed-id-updates/) prior to C#9 and also reference [Thomas Levesque's approach](https://thomaslevesque.com/2021/03/19/csharp-9-records-as-strongly-typed-ids-part-5-final-bits-and-conclusion/) using records [in this post](https://andrewlock.net/rebuilding-stongly-typed-id-as-a-source-generator-1-0-0-beta-release/).

The reality is though, if you want _nice_ serialization, then Records alone won't solve it for you, so you'll probably want some sort of source generator
